src/ucp/wireup/wireup_ep.c

```diff
@@ -52,11 +52,70 @@ static ssize_t ucp_wireup_ep_bcopy_send_func(uct_ep_h uct_ep)
     return UCS_ERR_NO_RESOURCE;
 }
 
+static void
+ucp_wireup_ep_pending_req_release(uct_pending_req_t *self, void *arg)
+{
+    ucp_request_t   *proxy_req = ucs_container_of(self, ucp_request_t,
+                                                  send.uct);
+    ucp_wireup_ep_t *wireup_ep = proxy_req->send.proxy.wireup_ep;
+    ucp_request_t   *req;
+
+    ucs_atomic_sub32(&wireup_ep->pending_count, 1);
+
+    if (proxy_req->send.proxy.req->func == ucp_wireup_msg_progress) {
+        req = ucs_container_of(proxy_req->send.proxy.req, ucp_request_t,
+                               send.uct);
+        ucs_free(req->send.buffer);
+        ucp_request_mem_free(req);
+    }
+
+    ucs_free(proxy_req);
+}
+
+static void
+ucp_wireup_ep_pending_purge(uct_ep_h uct_ep, uct_pending_purge_callback_t cb,
+                            void *arg)
+{
+    ucp_wireup_ep_t *wireup_ep = ucp_wireup_ep(uct_ep);
+
+    ucp_wireup_ep_pending_queue_purge(uct_ep, cb, arg);
+
+    if (wireup_ep->pending_count > 0) {
+        uct_ep_pending_purge(ucp_wireup_ep_get_msg_ep(wireup_ep),
+                             ucp_wireup_ep_pending_req_release, arg);
+    }
+
+    ucs_assert(wireup_ep->pending_count == 0);
+}
+
+static int ucp_wireup_ep_is_next_ep_active(ucp_wireup_ep_t *wireup_ep)
+{
+    return (wireup_ep->flags & UCP_WIREUP_EP_FLAG_READY) ||
+           (wireup_ep->aux_ep == NULL);
+}
+
+uct_ep_h ucp_wireup_ep_extract_msg_ep(ucp_wireup_ep_t *wireup_ep)
+{
+    uct_ep_h msg_ep = ucp_wireup_ep_get_msg_ep(wireup_ep);
+
+    ucp_wireup_ep_pending_purge(&wireup_ep->super.super,
+                                ucs_empty_function_do_assert_void, NULL);
+
+    if (ucp_wireup_ep_is_next_ep_active(wireup_ep)) {
+        ucp_wireup_ep_extract_next_ep(&wireup_ep->super.super);
+    } else {
+        wireup_ep->aux_ep        = NULL;
+        wireup_ep->aux_rsc_index = UCP_NULL_RESOURCE;
+    }
+
+    return msg_ep;
```

Comment:
don't use the function ucp_wireup_ep_get_msg_ep
and then:

```suggestion
    if (ucp_wireup_ep_is_next_ep_active(wireup_ep)) {
        return ucp_wireup_ep_extract_next_ep(&wireup_ep->super.super);
    }
    
    msg_ep                   = wireup->aux_ep;
    wireup_ep->aux_ep        = NULL;
    wireup_ep->aux_rsc_index = UCP_NULL_RESOURCE;
    return msg_ep;
```